src/python/pants/base/exception_sink.py

```diff
@@ -59,20 +268,101 @@ def _format_exception_message(cls, msg, pid):
+    cls._exiter.exit(result=cls.UNHANDLED_EXCEPTION_EXIT_CODE,
+                     msg=terminal_msg,
+                     out=cls._interactive_output_stream)
+
+  @classmethod
+  def _log_unhandled_exception_and_exit(cls, exc_class=None, exc=None, tb=None, add_newline=False):
+    """Default sys.excepthook implementation for unhandled exceptions."""
+    exc_class = exc_class or sys.exc_info()[0]
+    exc = exc or sys.exc_info()[1]
+    tb = tb or sys.exc_info()[2]
+
+    extra_err_msg = None
     try:
-      pid = os.getpid()
-      fatal_error_log_entry = cls._format_exception_message(msg, pid)
-      # We care more about this log than the shared log, so completely write to it first. This
-      # avoids any errors with concurrent modification of the shared log affecting the per-pid log.
-      safe_file_dump(cls.exceptions_log_path(for_pid=pid), fatal_error_log_entry, mode='w')
-      # TODO: we should probably guard this against concurrent modification somehow.
-      safe_file_dump(cls.exceptions_log_path(), fatal_error_log_entry, mode='a')
+      # Always output the unhandled exception details into a log file, including the traceback.
+      exception_log_entry = cls._format_unhandled_exception_log(exc, tb, add_newline,
+                                                              should_print_backtrace=True)
+      cls.log_exception(exception_log_entry)
     except Exception as e:
-      # TODO: If there is an error in writing to the exceptions log, we may want to consider trying
-      # to write to another location (e.g. the cwd, if that is not already the destination).
-      logger.error('Problem logging original exception: {}'.format(e))
+      extra_err_msg = 'Additional error logging unhandled exception {}: {}'.format(exc, e)
+      logger.error(extra_err_msg)
+
+    # Generate an unhandled exception report fit to be printed to the terminal (respecting the
+    # Exiter's _should_print_backtrace field).
+    # TODO: the Exiter just prints to stderr -- is that always open? Should we manage that here too?
+    stderr_printed_error = cls._format_unhandled_exception_log(
+      exc, tb, add_newline,
+      should_print_backtrace=cls._exiter.should_print_backtrace)
+    if extra_err_msg:
+      stderr_printed_error = '{}\n{}'.format(stderr_printed_error, extra_err_msg)
+    cls._exit_with_failure(stderr_printed_error)
+
+  _CATCHABLE_SIGNAL_ERROR_LOG_FORMAT = """\
+Signal {signum} was raised. Exiting with failure.
+{formatted_traceback}
+"""
+
+  _is_handling_signal = False
+
+  @classmethod
+  def handle_signal_gracefully(cls, signum, frame):
+    if signum in cls.KEYBOARD_INTERRUPT_SIGNALS:
+      raise KeyboardInterrupt('User interrupted execution with control-c!')
+    # TODO: determine the appropriate signal-safe behavior here (to avoid writing to our file
+    # descriptors re-entrantly, which raises an IOError). This is a best-effort solution.
+    if cls._is_handling_signal:
```

Comment:
a docstring here to explain what signals this is actually handling, would be great.

Reply:
Will do!

Reply:
(done!)